transports/quic/src/socket.rs

```diff
@@ -0,0 +1,138 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! I/O for libp2p-quic.
+//!
+//! This provides a central location for socket I/O, and logs all incoming and outgoing packets.
+
+use async_std::net::UdpSocket;
+use log::{trace, warn};
+use quinn_proto::Transmit;
+use std::{future::Future, io::Result, task::Context, task::Poll};
+
+/// A pending packet for libp2p-quic
+#[derive(Debug, Default)]
+pub(crate) struct Pending {
+    pending: Option<Transmit>,
+}
+
+/// A socket wrapper for libp2p-quic
+#[derive(Debug)]
+pub(crate) struct Socket {
+    socket: UdpSocket,
+}
+
+impl Socket {
+    /// Transmit a packet if possible, with appropriate logging.
+	///
+	/// We ignore I/O errors.  If a packet cannot be sent, we assume it is a transient condition
+	/// and drop it.  If it is not, the connection will eventually time out.  This provides a very
+	/// high degree of robustness.  Connections will transparently resume after a transient network
+	/// outage, and problems that are specific to one peer will not effect other peers.
+    pub fn poll_send_to(&self, cx: &mut Context, packet: &Transmit) -> Poll<Result<()>> {
+        match {
+            let fut = self.socket.send_to(&packet.contents, &packet.destination);
+            futures::pin_mut!(fut);
+            fut.poll(cx)
+        } {
+            Poll::Pending => {
+                trace!("not able to send packet right away");
+                Poll::Pending
+            }
+            Poll::Ready(Ok(e)) => {
+                trace!("sent packet of length {} to {}", e, packet.destination);
+                Poll::Ready(Ok(()))
+            }
+            Poll::Ready(Err(e)) => {
+                warn!("Ignoring I/O error on transmit: {:?}", e);
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    /// A wrapper around `recv_from` that handles ECONNRESET and logging
+    pub fn recv_from(
+        &self,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<Result<(usize, std::net::SocketAddr)>> {
+        loop {
+            match {
+                let fut = self.socket.recv_from(buf);
+                futures::pin_mut!(fut);
+                fut.poll(cx)
+            } {
+                Poll::Pending => {
+                    trace!("no packets available yet");
+                    break Poll::Pending;
+                }
+                Poll::Ready(Ok(e)) => {
+                    trace!("received packet of length {} from {}", e.0, e.1);
+                    break Poll::Ready(Ok(e));
+                }
+                // May be injected by a malicious ICMP packet
```

Comment:
Do you have info how that works?
I'm not too familiar with UDP, and I don't understand how a UDP socket could produce `ConnectionReset`.

Reply:
I *believe* it is ICMP host unreachable, but really am not sure.  I just know that quinn does the same thing.